raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -258,7 +260,8 @@ public KafkaRaftClient(
             localSupportedKRaftVersion,
             logContext,
             new Random(),
-            quorumConfig
+            quorumConfig,
+            new HashSet<>()
```

Comment:
KRaft should not track the starting offsets in this way. There are many issues with this approach. The more important ones are:

1. This doesn't track batch start offsets that were added to the log prior to starting the KRaft replica. It also doesn't track batch start offsets that were added by the leader.
2. This doesn't handle log truncation.
3. It is very memory inefficient. It is possible for KRaft to have a lot of small batches and each one of them would have an entry in this set.

The log layer already track these offsets. We shouldn't duplicate this information in the KRaft layer. Take a look at the `KafkaMetadataLog#read` method to see how the log layer computes the batch start offset. Specifically, you can take a look at `LogSegment#translateOffset`.

Reply:
Looks like the log layer gets the batch associated with a `targetOffset` from `FileRecords#searchForOffsetWithSize`, but it currently returns the last offset included in that batch. It looks safe to change this to return the base offset instead, so we can retrieve the batch start offset for a given offset. `LogSegment#read` also needs to construct `LogOffsetMetadata` with this batch start offset instead of the `startOffset` that is passed into the method.